### PR TITLE
Create site key on publisher creation

### PR DESCRIPTION
**Pull request by mcollins-ttd**

### CODE REVIEW

File: src/main/java/com/uid2/admin/secret/IEncryptionKeyManager.java [hunk 1]
```diff
@@ -4,4 +4,9 @@
 
 public interface IEncryptionKeyManager {
     EncryptionKey addSiteKey(int siteId) throws Exception;
+
+    /**
+     * Creates a site key, if none exists. If created, the key is active immediately.
+     */
+    EncryptionKey ensureSiteKeyExists(int siteId) throws Exception;
```
Comment (jon8787) [inline src/main/java/com/uid2/admin/secret/IEncryptionKeyManager.java:11]: I've seen "ensure" used to mean "check or throw"
Maybe something like "getOrCreateSiteKey"?

File: src/main/java/com/uid2/admin/vertx/service/ClientKeyService.java [hunk 2]
```diff
@@ -414,10 +422,18 @@ private void handleClientRoles(RoutingContext rc) {
...
             // return client with new key
             rc.response().end(jsonWriter.writeValueAsString(c));
         } catch (Exception e) {
             rc.fail(500, e);
         }
     }
+
+    private void ensurePublisherSiteKeyExists(ClientKey clientKey) throws Exception {
+        if (clientKey.hasRole(Role.GENERATOR)) {
+            this.encryptionKeyManager.ensureSiteKeyExists(clientKey.getSiteId());
```
Comment (jon8787) [inline src/main/java/com/uid2/admin/vertx/service/ClientKeyService.java:436]: is the return value ever used or could ensureSiteKeyExists return void?

File: src/main/java/com/uid2/admin/vertx/service/ClientKeyService.java [hunk 3]
```diff
@@ -414,10 +422,18 @@ private void handleClientRoles(RoutingContext rc) {
...
+
             // return client with new key
             rc.response().end(jsonWriter.writeValueAsString(c));
         } catch (Exception e) {
             rc.fail(500, e);
         }
     }
+
+    private void ensurePublisherSiteKeyExists(ClientKey clientKey) throws Exception {
+        if (clientKey.hasRole(Role.GENERATOR)) {
```
Comment (jon8787) [inline src/main/java/com/uid2/admin/vertx/service/ClientKeyService.java:435]: actually Role.SHARER should also have site keys

File: src/main/java/com/uid2/admin/vertx/service/ClientKeyService.java [hunk 4]
```diff
@@ -414,10 +422,18 @@ private void handleClientRoles(RoutingContext rc) {
             // upload to storage
             storeWriter.upload(clients, null);
 
+            ensurePublisherSiteKeyExists(c);
```
Comment (jon8787) [inline src/main/java/com/uid2/admin/vertx/service/ClientKeyService.java:425]: I wonder if there's any deduplication we could do here. eg add & update  both upload/ensure/respond, and future changes to these would happen in sync

File: src/main/java/com/uid2/admin/secret/IEncryptionKeyManager.java (hunk 1 above)
Comment (mcollins-ttd) [reply src/main/java/com/uid2/admin/secret/IEncryptionKeyManager.java]: 👍 Renamed to `createSiteKeyIfNoneExists`

File: src/main/java/com/uid2/admin/vertx/service/ClientKeyService.java (hunk 3 above)
Comment (mcollins-ttd) [reply src/main/java/com/uid2/admin/vertx/service/ClientKeyService.java:435]: Leaving for now, but let's come back to it when we have settled on the new roles.

**Review by jon8787 — APPROVED** [review]

**Pull request merged**